### PR TITLE
NSFS | NC | bucket and account manage CLI

### PR DESCRIPTION
**Pull request by naveenpaul1**
### Explain the changes
1. Users can create, update, delete, and list buckets and accounts using CLI 

### Issues: Fixed #xxx / Gap #xxx
1. https://issues.redhat.com/browse/MCGI-197

### Testing Instructions:
1. Please refer the doc : non_containerized_NSFS.md

- [ ] Doc added/updated
- [ ] Tests added

### CODE REVIEW

File: src/cmd/manage_nsfs.js [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+    try {
+        const resourse_type = argv._[0] || '';
+        if ((argv.help || argv.h) && resourse_type === 'account') return print_account_usage();
+        if ((argv.help || argv.h) && resourse_type === 'bucket') return print_bucket_usage();
+        if (argv.debug) {
+            const debug_level = Number(argv.debug) || 5;
+            dbg.set_module_level(debug_level, 'core');
+            nb_native().fs.set_debug_level(debug_level);
+        }
+        const config_root = argv.config_root ? String(argv.config_root) : '';
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:108]:
```suggestion
        const config_root = argv.config_root ? String(argv.config_root) : config.NSFS_NC_DEFAULT_CONF_DIR;
```

Comment (romayalon) [reply src/cmd/manage_nsfs.js:108]: it's not in master yet but I added /etc/noobaa.conf.d/ as config.NSFS_NC_DEFAULT_CONF_DIR

File: src/cmd/manage_nsfs.js [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+            console.log("Bucket exists with access key : " + data.name);
+        } else {
+            console.log("Bucket do not exists with access key : " + data.name);
+        }
+    } else if (action === 'update') {
+        const is_valid = await validate_bucket_add_args(data);
+        if (!is_valid) {
+            print_bucket_usage();
+        }
+        create_resourse(data.name, bucket_config_path, data);
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:209]: why do we create the resource on update? also I saw the on fetch we update the file

File: src/cmd/manage_nsfs.js [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        data.name = new SensitiveString(data.name);
+    }
+    if (!data) {
+        const bucket_path = argv.path ? String(argv.path) : '';
+        data = {
+            name: name,
+            system_owner: email,
+            bucket_owner: email,
+            tag: "",
+            versioning: "DISABLED",
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:156]: let's replace all double quotes (") to single quotes (')

File: src/cmd/manage_nsfs.js [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+        print_account_usage();
+    }
+    const data = _.merge({}, source, target);
+    return data;
+}
+
+async function manage_bucket_operations(action, data, config_root) {
+    const bucket_config_path = config_root + '/buckets';
+    if (action === 'add') {
+        const is_valid = await validate_bucket_add_args(data);
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:186]: I would suggest to split this function to multiple functions, each handling an action

File: src/cmd/manage_nsfs.js [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+    const action = argv._[1] || '';
+    const config_root = argv.config_root ? String(argv.config_root) : '';
+    const data = await fetch_account_data(argv);
+    manage_account_operations(action, data, config_root);
+}
+
+async function fetch_account_data(argv) {
+    let data;
+    const action = argv._[1] || '';
+    const from_file = argv.from_file ? String(argv.from_file) : '';
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:234]: it looks like from_file and config_root assignment can be done once in main, WDYT? 
const config_root = argv.config_root ? String(argv.config_root) : ''; is done many times and I don't think it's necassary

File: src/cmd/manage_nsfs.js [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+    } else if (action === 'list') {
+        const accounts = await list_resources(account_path);
+        console.log(accounts);
+    } else {
+        console.log("Account action not found.");
+        print_account_usage();
+    }
+}
+
+async function create_resourse(file_name, file_path, data) {
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:333]: I would suggest to change the create_resource to create_config_file(), resource has a different meaning in NooBaa and I find it a bit confusing

File: src/cmd/manage_nsfs.js (hunk 2 above)
Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js:209]: @romayalon its calling method `await fs_utils.replace_file(full_path, data);`. It creates the file if the file missing and replaces already exists. On fetch we we are just getting data from the config file and updating object with input data.

File: src/cmd/manage_nsfs.js [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+    # Read Bucket details from json file, no need to mention all the propery one by one in CLI
+    --from_file <dir>           (default none)          Set buckt/account schema full path.
+                                                        Get bucket and account details from json file
+    # required for add, replace 
+    --email <email>             (default none)          Set the email for account/bucket.
+    --path <dir>                (default none)          Set the bucket path.
+
+    # required for add, replace and delete
+    --name <name>               (default none)          Set the name for account/bucket.
+    --config_root <dir>         (default none)          Configuration files path for Noobaa standalon NSFS.
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:69]:
```suggestion
    --config_root <dir>         (default config.NSFS_NC_DEFAULT_CONF_DIR)          Configuration files path for Noobaa standalon NSFS.
```

File: src/cmd/manage_nsfs.js [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+    }
+    return false;
+}
+
+async function main(argv = minimist(process.argv.slice(2))) {
+    try {
+        const resourse_type = argv._[0] || '';
+        if ((argv.help || argv.h) && resourse_type === 'account') return print_account_usage();
+        if ((argv.help || argv.h) && resourse_type === 'bucket') return print_bucket_usage();
+        if (argv.debug) {
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:104]: so, do we support updating the debug level through this CLI? might be a good idea, we just need it documented and tested

File: src/cmd/manage_nsfs.js [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+
+async function get_bucket_config_file_status(data, bucket_config_path) {
+    const is_valid = await validate_minimum_bucket_args(data);
+    if (!is_valid) {
+        print_account_usage();
+        return;
+    }
+    const bucket_dir_stat = await fs_utils.file_exists(path.join(bucket_config_path, data.name + '.json'));
+    if (bucket_dir_stat) {
+        console.log('Bucket exists with access key : ' + data.name);
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:198]: why not print all the bucket data?

File: src/cmd/manage_nsfs.js [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+    const full_path = path.join(file_path, file_name.unwrap() + '.json');
+    await fs_utils.replace_file(full_path, data);
+    console.log('Resourse created : ', file_name.unwrap());
+}
+
+async function update_config_file(file_name, file_path, data) {
+    data = JSON.stringify(data);
+    const full_path = path.join(file_path, file_name.unwrap() + '.json');
+    await fs_utils.replace_file(full_path, data);
+    console.log('Resourse updated : ', file_name.unwrap());
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:373]:
```suggestion
    console.log('Resource updated : ', file_name.unwrap());
```

Comment (romayalon) [reply src/cmd/manage_nsfs.js:373]: please replace it all over the file

File: src/cmd/manage_nsfs.js [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+    --new_buckets_path <dir>    (default none)          Set the root of the filesystem where each subdir is a bucket.
+    
+    # required for add, replace and delete
+    --access_key <key>          (default none)          Authenticate incoming requests for this access key only (default is no auth).
+    --config_root <dir>         (default none)          Configuration files path for Noobaa standalon NSFS.
+`;
+
+const BUCKET_OPTIONS = `
+Bucket Options:
+
```
Comment (romayalon) [inline src/cmd/manage_nsfs.js:59]: I think it's important here to remind the user that create bucket cli command won't create the directory and the bucket directory should exist prior to the create bucket cli command

Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js:59]: @romayalon but we are creating the bucket for the user from cli in method `check_and_create_config_dirs()`, Should I remove this?

File: src/cmd/manage_nsfs.js (hunk 8 above)
Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js:104]: @romayalon When I thought about it, it doesn't look good to have debug level changes in this CLI. WDYT? Should I remove it?

File: src/cmd/manage_nsfs.js (hunk 11 above)
Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js:59]: @romayalon oh.. ok, I got confused it with config_root dir, Yes I will mention about bucke dir in doc, Thank you

File: src/cmd/manage_nsfs.js (hunk 8 above)
Comment (romayalon) [reply src/cmd/manage_nsfs.js]: you can remove it from this PR and we should think how to let users do that

**Review by romayalon — APPROVED** [review]

File: src/cmd/manage_nsfs.js (hunk 8 above)
Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js]: done

**Pull request merged**